File: latest/week2/workshop6_m3/W06_M03_demo.py
"""W06-M03: Line Charts in Code — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.dates import DateFormatter, MonthLocator, YearLocator
from matplotlib.gridspec import GridSpec
os.makedirs("output", exist_ok=True); np.random.seed(42)

months = pd.date_range("2020-01", "2023-12", freq="ME")
n = len(months)
val = 100 + np.cumsum(np.random.normal(0.5, 3, n))

# ── 1. ax.plot(): BASIC LINE CHART ──────────────────────
fig, ax = plt.subplots(figsize=(9, 4))
ax.plot(months, val, color="#1565C0", lw=1.5)
ax.xaxis.set_major_formatter(DateFormatter("%b %Y"))
ax.xaxis.set_major_locator(MonthLocator(interval=6))
plt.setp(ax.get_xticklabels(), rotation=30, fontsize=7)
ax.set_title("ax.plot(): Continuous Time Series\n"
    "DateFormatter('%b %Y') + MonthLocator(interval=6)", fontweight="bold")
ax.set_ylabel("Value"); ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/line_basic.png", dpi=300); plt.close()

# ── 2. ax.step(): DISCRETE CHANGES ──────────────────────
rates = np.array([5.5]*12 + [4.5]*12 + [3.0]*12 + [3.5]*12)
change_dates = [pd.Timestamp("2021-01-01"), pd.Timestamp("2022-01-01"),
    pd.Timestamp("2023-01-01")]
change_rates = [4.5, 3.0, 3.5]

fig, ax = plt.subplots(figsize=(8, 4))
ax.step(months, rates, color="#E53935", lw=1.5, where="post")
ax.scatter(change_dates, change_rates, color="#E53935", s=40, zorder=5,
    edgecolors="white", lw=1)
ax.xaxis.set_major_formatter(DateFormatter("%Y"))
ax.set_title("ax.step(): Discrete Rate Changes\n"
    "Horizontal segments + vertical jumps", fontweight="bold")
ax.set_ylabel("Interest Rate (%)"); ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/step_chart.png", dpi=300); plt.close()

# ── 3. fill_between(): AREA CHART ───────────────────────
fig, ax = plt.subplots(figsize=(8, 4))
ax.fill_between(months, val, alpha=0.3, color="#1565C0")
ax.plot(months, val, color="#1565C0", lw=0.8)
ax.xaxis.set_major_formatter(DateFormatter("%b\n%Y"))
ax.xaxis.set_major_locator(MonthLocator(interval=6))
ax.set_title("fill_between(): Filled Area\n"
    "Emphasises magnitude / cumulative volume", fontweight="bold")
ax.set_ylabel("Value"); ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/area_chart.png", dpi=300); plt.close()

# ── 4. stackplot(): STACKED AREA ────────────────────────
v1 = np.abs(np.random.normal(30, 5, n))
v2 = np.abs(np.random.normal(25, 5, n))
v3 = np.abs(np.random.normal(15, 5, n))

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 5))

# Absolute stacked
ax1.stackplot(months, v1, v2, v3, labels=["Product A", "Product B", "Product C"],
    colors=["#1565C0", "#E53935", "#2E7D32"], alpha=0.7)
ax1.legend(fontsize=7, loc="upper left")
ax1.set_title("Stacked Area: Absolute Values\n"
    "Bands sum to total", fontsize=10, fontweight="bold")
ax1.set_ylabel("Revenue (EUR)")
ax1.xaxis.set_major_formatter(DateFormatter("%Y"))

# 100% stacked
total = v1 + v2 + v3
ax2.stackplot(months, v1/total*100, v2/total*100, v3/total*100,
    labels=["Product A", "Product B", "Product C"],
    colors=["#1565C0", "#E53935", "#2E7D32"], alpha=0.7)
ax2.legend(fontsize=7, loc="upper left")
ax2.set_title("100% Stacked: Proportions\n"
    "Each point sums to 100%", fontsize=10, fontweight="bold")
ax2.set_ylabel("Share (%)")
ax2.xaxis.set_major_formatter(DateFormatter("%Y"))

for ax in [ax1, ax2]:
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/stacked_area.png", dpi=300); plt.close()

# ── 5. fill_between(): RIBBON (CONFIDENCE BAND) ─────────
lower = val - 8 - np.random.uniform(0, 3, n)
upper = val + 8 + np.random.uniform(0, 3, n)

fig, ax = plt.subplots(figsize=(8, 4))
ax.fill_between(months, lower, upper, alpha=0.15, color="#1565C0")
ax.plot(months, val, color="#1565C0", lw=1.5)
ax.set_title("fill_between(lower, upper): Confidence Ribbon\n"
    "Shaded region = uncertainty around central estimate", fontweight="bold")
ax.set_ylabel("Value"); ax.xaxis.set_major_formatter(DateFormatter("%b %Y"))
ax.xaxis.set_major_locator(MonthLocator(interval=6))
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/ribbon.png", dpi=300); plt.close()

# ── 6. SIX-PANEL COMPARISON DASHBOARD ───────────────────
fig = plt.figure(figsize=(14, 8))
gs = GridSpec(2, 3, figure=fig, hspace=0.4, wspace=0.3)

ax = fig.add_subplot(gs[0,0])
ax.plot(months, val, color="#1565C0", lw=1.2)
ax.set_title("(a) Line: ax.plot()", fontsize=9, fontweight="bold")

ax = fig.add_subplot(gs[0,1])
ax.step(months, rates, color="#E53935", lw=1.2, where="post")
ax.set_title("(b) Step: ax.step()", fontsize=9, fontweight="bold")

ax = fig.add_subplot(gs[0,2])
ax.fill_between(months, val, alpha=0.3, color="#2E7D32")
ax.plot(months, val, color="#2E7D32", lw=0.5)
ax.set_title("(c) Area: fill_between()", fontsize=9, fontweight="bold")

ax = fig.add_subplot(gs[1,0])
ax.stackplot(months, v1, v2, v3, colors=["#1565C0","#E53935","#2E7D32"], alpha=0.7)
ax.set_title("(d) Stacked: stackplot()", fontsize=9, fontweight="bold")

ax = fig.add_subplot(gs[1,1])
ax.stackplot(months, v1/total*100, v2/total*100, v3/total*100,
    colors=["#1565C0","#E53935","#2E7D32"], alpha=0.7)
ax.set_title("(e) 100% Stacked", fontsize=9, fontweight="bold")

ax = fig.add_subplot(gs[1,2])
ax.fill_between(months, lower, upper, alpha=0.15, color="#1565C0")
ax.plot(months, val, color="#1565C0", lw=1.2)
ax.set_title("(f) Ribbon: fill_between(CI)", fontsize=9, fontweight="bold")

for a in fig.get_axes():
    a.tick_params(labelsize=5); a.spines["top"].set_visible(False)
    a.spines["right"].set_visible(False)
    a.xaxis.set_major_formatter(DateFormatter("%Y"))
fig.suptitle("Six Line Geometries in Python", fontsize=13, fontweight="bold")
plt.savefig("output/six_geoms.png", dpi=300, bbox_inches="tight"); plt.close()

# ── 7. DATE FORMATTING COMPARISON ───────────────────────
fig, axes = plt.subplots(2, 2, figsize=(12, 7))
fmts = [("%b %Y", "Jan 2020"), ("%Y-%m", "2020-01"),
        ("%Y", "2020"), ("%b\\n%Y", "Jan\\n2020")]
for ax, (fmt, desc) in zip(axes.flatten(), fmts):
    ax.plot(months[:24], val[:24], color="#1565C0", lw=1.2)
    ax.xaxis.set_major_formatter(DateFormatter(fmt.replace("\\n", "\n")))
    ax.xaxis.set_major_locator(MonthLocator(interval=3))
    ax.set_title(f"Format: '{fmt}' → {desc}", fontsize=9, fontweight="bold")
    ax.tick_params(labelsize=6)
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Date Formatting Options in matplotlib", fontsize=12, fontweight="bold")
plt.tight_layout(); plt.savefig("output/date_formats.png", dpi=300); plt.close()

# ── 8. REAL DATA: NETFLIX ───────────────────────────────
nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf = nf.dropna(subset=["date_added"]).query("date_added.dt.year >= 2015")

# Monthly additions
nf_monthly = nf.set_index("date_added").resample("ME").size()

fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 8), sharex=True)

# Line chart
ax1.plot(nf_monthly.index, nf_monthly.values, color="#E53935", lw=1)
ax1.set_title("Netflix: Monthly Additions (line)", fontweight="bold")
ax1.set_ylabel("Titles Added")

# Stacked area by type
type_monthly = nf.set_index("date_added").groupby("type").resample("ME").size().unstack(0, fill_value=0)
if "Movie" in type_monthly and "TV Show" in type_monthly:
    ax2.stackplot(type_monthly.index,
        type_monthly["Movie"], type_monthly["TV Show"],
        labels=["Movie", "TV Show"],
        colors=["#1565C0", "#E53935"], alpha=0.7)
    ax2.legend(fontsize=7)
ax2.set_title("Netflix: Monthly Additions by Type (stacked area)", fontweight="bold")
ax2.set_ylabel("Titles Added")
ax2.xaxis.set_major_formatter(DateFormatter("%b %Y"))

for ax in [ax1, ax2]:
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/netflix_lines.png", dpi=300); plt.close()

# ── CHEATSHEET ───────────────────────────────────────────
print("\n=== Line Geometry Cheatsheet ===")
print("ax.plot(x, y)          → continuous line (default)")
print("ax.step(x, y, where=)  → discrete jumps ('pre','post','mid')")
print("ax.fill_between(x, y)  → filled area below line")
print("ax.stackplot(x,y1,y2)  → stacked area (parts of whole)")
print("ax.fill_between(x,lo,hi) → ribbon (confidence band)")
print("")
print("=== Date Formatting ===")
print("DateFormatter('%b %Y')  → Jan 2020")
print("DateFormatter('%Y-%m')  → 2020-01")
print("MonthLocator(interval=3) → tick every 3 months")
print("YearLocator()            → tick every year")

print("\nAll W06-M03 Python outputs saved")

File: latest/week2/workshop6_m3/W06_M03_demo.R
# ============================================================
# Workshop 6 — Module 3: Line Charts in Code (R + Python)
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── 1. SIMULATED DATA ───────────────────────────────────
set.seed(42)
months <- seq(as.Date("2020-01-01"), as.Date("2023-12-01"), by = "month")
n <- length(months)
val <- 100 + cumsum(rnorm(n, 0.5, 3))

# ── 2. geom_line: CONTINUOUS TIME SERIES ─────────────────
# Basic line chart with proper date axis formatting
p_line <- ggplot(tibble(date = months, value = val), aes(x = date, y = value)) +
  geom_line(color = "#1565C0", linewidth = 1) +
  scale_x_date(
    date_breaks = "6 months",
    date_labels = "%b %Y",        # abbreviated month + 4-digit year
    date_minor_breaks = "1 month") +  # minor gridlines every month
  theme_minimal() +
  labs(title = "geom_line: Continuous Time Series",
       subtitle = "date_breaks = '6 months', date_labels = '%b %Y'",
       x = NULL, y = "Value")
ggsave("output/line_basic.png", p_line, width = 9, height = 4, dpi = 300)

# ── 3. geom_step: DISCRETE CHANGES ──────────────────────
# Interest rate changes (discrete jumps, constant between)
rates <- c(rep(5.5, 12), rep(4.5, 12), rep(3.0, 12), rep(3.5, 12))
p_step <- ggplot(tibble(date = months, rate = rates), aes(x = date, y = rate)) +
  geom_step(color = "#E53935", linewidth = 1.5) +
  geom_point(data = tibble(
    date = as.Date(c("2021-01-01", "2022-01-01", "2023-01-01")),
    rate = c(4.5, 3.0, 3.5)),
    color = "#E53935", size = 3) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  theme_minimal() +
  labs(title = "geom_step: Discrete Rate Changes",
       subtitle = "Value jumps at specific dates, constant between",
       x = NULL, y = "Interest Rate (%)")
ggsave("output/step_chart.png", p_step, width = 8, height = 4, dpi = 300)

# ── 4. geom_area: FILLED MAGNITUDE ──────────────────────
p_area <- ggplot(tibble(date = months, value = val), aes(x = date, y = value)) +
  geom_area(fill = "#1565C0", alpha = 0.3, color = "#1565C0", linewidth = 0.5) +
  scale_x_date(date_breaks = "6 months", date_labels = "%b\n%Y") +
  theme_minimal() +
  labs(title = "geom_area: Filled Magnitude",
       subtitle = "Like geom_line but fills below — emphasises cumulative volume",
       x = NULL, y = "Value")
ggsave("output/area_chart.png", p_area, width = 8, height = 4, dpi = 300)

# ── 5. STACKED AREA: PARTS OF A WHOLE OVER TIME ─────────
v1 <- abs(rnorm(n, 30, 5)); v2 <- abs(rnorm(n, 25, 5)); v3 <- abs(rnorm(n, 15, 5))
df_stack <- tibble(
  date = rep(months, 3),
  value = c(v1, v2, v3),
  product = rep(c("Product A", "Product B", "Product C"), each = n))

# Absolute stacked area
p_stacked <- ggplot(df_stack, aes(x = date, y = value, fill = product)) +
  geom_area(alpha = 0.7, color = "white", linewidth = 0.3) +
  scale_fill_manual(values = c("#1565C0", "#E53935", "#2E7D32")) +
  theme_minimal() +
  labs(title = "Stacked Area: Parts of a Whole Over Time",
       subtitle = "Each band = one product; total = sum of all bands",
       x = NULL, y = "Revenue (EUR)")
ggsave("output/stacked_area.png", p_stacked, width = 9, height = 5, dpi = 300)

# 100% stacked area (proportions)
df_pct <- df_stack |>
  group_by(date) |>
  mutate(pct = value / sum(value) * 100) |>
  ungroup()

p_pct <- ggplot(df_pct, aes(x = date, y = pct, fill = product)) +
  geom_area(alpha = 0.7, color = "white", linewidth = 0.3) +
  scale_fill_manual(values = c("#1565C0", "#E53935", "#2E7D32")) +
  scale_y_continuous(labels = function(x) paste0(x, "%")) +
  theme_minimal() +
  labs(title = "100% Stacked Area: Proportional Composition Over Time",
       subtitle = "Each point sums to 100% — shows relative share, not absolute",
       x = NULL, y = "Share (%)")
ggsave("output/stacked_100pct.png", p_pct, width = 9, height = 5, dpi = 300)

# ── 6. RIBBON: CONFIDENCE BAND ──────────────────────────
df_ribbon <- tibble(
  date = months,
  mean = val,
  lower = val - 8 - runif(n, 0, 3),
  upper = val + 8 + runif(n, 0, 3))

p_ribbon <- ggplot(df_ribbon, aes(x = date)) +
  geom_ribbon(aes(ymin = lower, ymax = upper),
    fill = "#1565C0", alpha = 0.15) +
  geom_line(aes(y = mean), color = "#1565C0", linewidth = 1) +
  theme_minimal() +
  labs(title = "geom_ribbon: Mean ± Confidence Band",
       subtitle = "Shaded region communicates uncertainty around the central estimate",
       x = NULL, y = "Value")
ggsave("output/ribbon.png", p_ribbon, width = 8, height = 4, dpi = 300)

# ── 7. SIX-PANEL COMPARISON DASHBOARD ───────────────────
p1 <- ggplot(tibble(date=months,value=val),aes(date,value)) +
  geom_line(color="#1565C0",linewidth=0.8) + theme_minimal(base_size=7) +
  labs(title="(a) Line")
p2 <- ggplot(tibble(date=months,rate=rates),aes(date,rate)) +
  geom_step(color="#E53935",linewidth=0.8) + theme_minimal(base_size=7) +
  labs(title="(b) Step")
p3 <- ggplot(tibble(date=months,value=val),aes(date,value)) +
  geom_area(fill="#2E7D32",alpha=0.3)+geom_line(color="#2E7D32",linewidth=0.5) +
  theme_minimal(base_size=7) + labs(title="(c) Area")
p4 <- ggplot(df_stack,aes(date,value,fill=product)) +
  geom_area(alpha=0.7,color="white",linewidth=0.2) +
  scale_fill_manual(values=c("#1565C0","#E53935","#2E7D32")) +
  theme_minimal(base_size=7) + theme(legend.position="none") +
  labs(title="(d) Stacked Area")
p5 <- ggplot(df_pct,aes(date,pct,fill=product)) +
  geom_area(alpha=0.7,color="white",linewidth=0.2) +
  scale_fill_manual(values=c("#1565C0","#E53935","#2E7D32")) +
  theme_minimal(base_size=7) + theme(legend.position="none") +
  labs(title="(e) 100% Stacked")
p6 <- ggplot(df_ribbon,aes(x=date)) +
  geom_ribbon(aes(ymin=lower,ymax=upper),fill="#1565C0",alpha=0.15) +
  geom_line(aes(y=mean),color="#1565C0",linewidth=0.8) +
  theme_minimal(base_size=7) + labs(title="(f) Ribbon (CI)")

dashboard <- (p1 | p2 | p3) / (p4 | p5 | p6) +
  plot_annotation(title = "Six Line Geometries in R",
    subtitle = "Same simulated time series, different visual encodings",
    tag_levels = "a") &
  theme_minimal(base_size = 7)
ggsave("output/six_geoms_dashboard.png", dashboard, width = 14, height = 8, dpi = 300)

# ── 8. DATE FORMATTING EXAMPLES ──────────────────────────
p_fmt1 <- ggplot(tibble(date=months[1:24],value=val[1:24]),aes(date,value)) +
  geom_line(color="#1565C0") +
  scale_x_date(date_breaks="3 months",date_labels="%b %Y") +
  theme_minimal(base_size=8) + labs(title="Format: %b %Y → Jan 2020")

p_fmt2 <- ggplot(tibble(date=months[1:24],value=val[1:24]),aes(date,value)) +
  geom_line(color="#1565C0") +
  scale_x_date(date_breaks="3 months",date_labels="%Y-%m") +
  theme_minimal(base_size=8) + labs(title="Format: %Y-%m → 2020-01")

p_fmt3 <- ggplot(tibble(date=months[1:24],value=val[1:24]),aes(date,value)) +
  geom_line(color="#1565C0") +
  scale_x_date(date_breaks="1 year",date_labels="%Y") +
  theme_minimal(base_size=8) + labs(title="Format: %Y → 2020")

p_fmt4 <- ggplot(tibble(date=months[1:24],value=val[1:24]),aes(date,value)) +
  geom_line(color="#1565C0") +
  scale_x_date(date_breaks="3 months",date_labels="%b\n%Y") +
  theme_minimal(base_size=8) + labs(title="Format: %b\\n%Y (two lines)")

ggsave("output/date_formats.png", (p_fmt1|p_fmt2)/(p_fmt3|p_fmt4),
  width=12, height=7, dpi=300)

# ── 9. REAL DATA: NETFLIX ───────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added))) |>
  filter(!is.na(date_added), year(date_added) >= 2015)

# Monthly additions line chart
nf_monthly <- nf |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month)

p_nf_line <- ggplot(nf_monthly, aes(x = month, y = n)) +
  geom_line(color = "#E53935", linewidth = 0.8) +
  scale_x_date(date_breaks = "6 months", date_labels = "%b\n%Y") +
  theme_minimal() +
  labs(title = "Netflix: Monthly Additions (2015–2021)",
       x = NULL, y = "Titles Added")

# Stacked area by type
nf_type_monthly <- nf |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month, type)

p_nf_stacked <- ggplot(nf_type_monthly, aes(x = month, y = n, fill = type)) +
  geom_area(alpha = 0.7, color = "white", linewidth = 0.3) +
  scale_fill_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  theme_minimal() +
  labs(title = "Netflix: Monthly Additions by Type (stacked)",
       x = NULL, y = "Titles Added")

ggsave("output/netflix_lines.png", p_nf_line / p_nf_stacked,
  width = 10, height = 8, dpi = 300)

cat("\n── All W06-M03 R outputs saved ──\n")
cat("Figures: line_basic, step_chart, area_chart, stacked_area,\n")
cat("  stacked_100pct, ribbon, six_geoms_dashboard, date_formats,\n")
cat("  netflix_lines\n")
